### PR TITLE
Add a rewrite_with_tactic hook

### PR DESCRIPTION
**Pull request by R1kM**
This PR proposes the addition of a `rewrite_with_tactic` hook to F*.

This hook works similarly to the current `with_tactic`: it is a syntactic marker that tags a term with a tactic.
But compared to `with_tactic tau assertion`, the purpose of `rewrite_with_tactic tau term` is not to prove an assertion, but to rewrite `term` using tactic `tau`.

Internally, the implementation of this hook is similar to other synthesis facilities:
A goal `term == ?u` is generated, which must be solved by `tau`.
Any remaining non-trivial goal after running `tau`, or not having solved `?u` leads to
an error.
If synthesis is successful, the initial `term` is replaced by `?u`, which has been fully resolved.

The traversal of a term is the same as for `with_tactic`, i.e. it follows a bottom-up approach. In particular, this ensures that if `rewrite_with_tactic` appears inside another `rewrite_with_tactic`, the inner call is resolved first.

### CODE REVIEW

File: src/tactics/FStar.Tactics.Hooks.fs [hunk 1]
```diff
@@ -115,6 +119,43 @@ let by_tactic_interp (pol:pol) (e:Env.env) (t:term) : tres =
...
+        // Create a new uvar that must be equal to the initial term
+        let uvtm, _, g_imp = Env.new_implicit_var_aux "rewrite_with_tactic RHS" tm.pos e typ Allow_untyped None in
+
+        let u = e.universe_of e typ in
+        // eq2 is squashed already, so it's in Type0
+        let goal = U.mk_squash U_zero (U.mk_eq2 u typ tm uvtm) in
+        let gs, _ = run_tactic_on_typ tactic.pos tm.pos tactic e goal in
+
+        // Ensure that rewriting did not leave goals
+        List.iter (fun g ->
```
Comment (aseemr) [inline src/tactics/FStar.Tactics.Hooks.fs:135]: As far as I understand, a user tactic can always drop all the goals using `FStar.Tactics.Builtins.set_goals`, so is this check adding any insurance? Why can't you just drop this loop, resolve_implicits, and then report if some remain unresolved?

**Comment by aseemr** [comment]
Hi Aymeric, this looks quite straightforward to me. Left one comment about checking prop goals that I did not understand. Thanks!

File: examples/tactics/RewriteTactic.fst [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+   limitations under the License.
+*)
+module RewriteTactic
+
+open FStar.Tactics
+
+let lem () : Lemma (1 == 2) = admit ()
+let tau () = apply_lemma (`lem)
+
+let test _ =
```
Comment (aseemr) [inline examples/tactics/RewriteTactic.fst:23]: Could you please also add a negative testcase where, e.g., the tactic leaves a uvar and then the rewrite fails?

File: src/tactics/FStar.Tactics.Hooks.fs (hunk 1 above)
Comment (R1kM) [reply src/tactics/FStar.Tactics.Hooks.fs:135]: That does sound easier. I'm just a bit confused about what goal I should run force_trivial_guard on if not the list of remaining goals.

File: examples/tactics/RewriteTactic.fst (hunk 2 above)
Comment (R1kM) [reply examples/tactics/RewriteTactic.fst]: Done

**Pull request merged**